### ISSUE TITLE
On doc load focus

### ISSUE THREAD

**Issue by andrewsprinz** [madebymany/sir-trevor-js#349]
Is there any way to stop ST stealing focus on page load?

**Comment by SgtOddball** [comment]
Thats a good point, I only noticed it does this yesterday (finished a page with some options above where Sir Trevor turns up and noticed this when the page gets reloaded). 

I might have a look around and see if I can spot where it's triggering the focus.

**Comment by SgtOddball** [comment]
This should prevent Sir Trev steal focus if you've not got any blocks but at the moment I think if there's text then the cursor gets moved to it. I'll keep digging to see If I can find a way round but for the time being you can add something like the following to your Sir Trev initialization script;

```javascript
    SirTrevor.setDefaults({
        onBlockRender: function (){
            console.log("Look ma, no focus.");
         }
    });
```

This is because it's overriding the default ```onBlockRender``` which triggers ```this.focus();```

**Comment by lukaszsagol** [comment]
I don't think you can override `onBlockRender` using `setDefaults` (unfortunately). It looks like overriding this function would require using `setBlockOptions` for every single block type.

It also means that SirTrevor won't automatically set focus for newly added blocks when editing.

**Comment by cjbell** [comment]
We should probably add a config option whether to set the initial focus / focus blocks once added. May have the implication that blocks need to know if they're the initial item though.

**Comment by SgtOddball** [comment]
As a thought, could there be an onload flag that gets unset once Sir trev has initialised for the focus? That way it wouldn't matter if the blocks are an initial item or not. 

Might look into this further If I get the chance (weather permitting :smile:)

**Comment by SgtOddball** [comment]
It started nagging me, so I've jury rigged a small script to execute where Sir Trev loads.

The gist being, it looks for any forms on a page, ignores hidden and disabled then picks the first input it finds in the dom. Once it's done that it makes sure the page has been scrolled to the top of the page (ymmv). 

All you need do is set this to fire on window.load with a small delay (I found 150ms to be sufficient) and it should happily set itself to the top of the page with the cursor patiently waiting at your first input.  

```javascript
    function formFocus(){
        var b = false;
        for( f = 0; f < document.forms.length ; f++ ){
            for( i = 0; i < document.forms[f].length; i++ ){
                if ( document.forms[f][i].type != "hidden" ){
                    if( document.forms[f][i].disabled != true ){
                        document.forms[f][i].focus();
                        var b=true
                    }
                }
                if( b == true ) break
            }
            if( b == true ) break
        };
        document.body.scrollTop = document.documentElement.scrollTop = 0;
    };
```

Also I know I could do it in jQuery but for once Javascript seemed the better bet overall.

**Comment by andrewsprinz** [comment]
Lovely hack, works a charm! :thumbsup:

**Comment by andrewsprinz** [comment]
wonder if there’s ST callback on instantiation or completion of render that this can be called from instead of a timeout?

**Comment by SgtOddball** [comment]
Maybe, but I haven't looked far enough into it.

When I've got abit of time I'm might have a route around and see if there's something that can be triggered after the initial load has finished. 

On saying.. had a brainstorm and I might have something

```javascript 
   SirTrevor.EventBus.on('block:create:existing',function(){
	var b = false;
        for( f = 0; f < document.forms.length ; f++ ){
            for( i = 0; i < document.forms[f].length; i++ ){
                if ( document.forms[f][i].type != "hidden" ){
                    if( document.forms[f][i].disabled != true ){
                        document.forms[f][i].focus();
                        var b=true
                    }
                }
                if( b == true ) break
            }
            if( b == true ) break
        };
        document.body.scrollTop = document.documentElement.scrollTop = 0;
   })
```

From a quick and dirty test, the ```block:create:existing``` event bus should only get called when loading up data already present in the textarea boxes. All you need do is fire the checks after Sir Trev has finished initialising and all being well it should be good to go. 

Only caveat I can think of is if you're adding data in after Sir Trev has loaded. I don't know yet if it still triggers the event or not.

**Comment by SgtOddball** [comment]
Did you get chance to try ```block:create:existing``` trigger yet?

**Comment by andrewsprinz** [comment]
tried it, should work in theory but I’m using an older release (0.4.3) which doesn’t trigger these events.  

from looking at that tag, the only way would be to modify the onBlockRender method, but no way to determine if it’s a new block or not

SirTrevor.setBlockOptions("Text", {
  onBlockRender: function() {
    console.log("Text block rendered");
  }
});


https://github.com/madebymany/sir-trevor-js/blob/b9691789037b4e1013609ffa214d0ab40760a622/examples-component/index.js

guess I can live with this knowing it’ll work in newer versions :)

**Comment by SgtOddball** [comment]
Ahh fare enough. Least you've got something in place for the time being.

**Comment by andrewsprinz** [comment]
So on using 0.5 I see this isn't so much an issue when rendering entirely empty Sir Trevor instances. The UI takes care of it elegantly without having to specify a default text block.

However it's still an issue when editing existing records. I wish I could contribute with a patch, but if I'm honest the tooling is a bit of a learning curve too far for me.

Seems the issue lies in block.js, line 191-ish.

`onBlockRender: function() {
    this.focus();
  },`

I understand that's it's important that this is maintained, but is there a way we could only fire this focus() if the block has not been rendered by :existing or by defining a default block? 

Maybe there's an event I don't know about which I could intercept?

**Comment by raffij** [comment]
@andrewsprinz this is something I was thinking about yesterday as I've got a long page with multiple sir-trevor instances on it and the jump isn't good.

`onBlockRender` making the focus happen doesn't feel like a good solution. It also makes it more difficult to override the behaviour on a global level.

I feel the desired behaviour is to move the focus logic to a function called from [BlockManager.js::createBlock] (https://github.com/madebymany/sir-trevor-js/blob/master/src/block-manager.js#L70)

The default logic would then be:

- Initial text block will focus depending on a global config setting.
- No focus for existing block rendering.
- When a new block is added focus.

This could then be overridden in the config. Feels better than an event which would fire and the internal function would do something and then the custom code would do something else.

I'll have a think.

**Comment by andrewsprinz** [comment]
That makes a whole lot of sense.

Will give it a try and report back.

On first attempt, I couldn’t get anything to listen to
`block:event:existing`, tried the element itself, and the global ST
eventbus using
` SirTrevor.EventBus.on(‘block:create:existing’,function(){`– is that still
the correct syntax?

**Comment by raffij** [comment]
@andrewsprinz looking at how embedded the focus functionality is within certain blocks such as the list block I think the easiest option is a fake blur using `onEditorRender`. Otherwise this could be done a block level using the 

Something similar to:

```js
onEditorRender: () => {
  let tempTextArea = document.createElement('div');
  tempTextArea.setAttribute('contenteditable', true);
  tempTextArea.style.position = 'absolute';
  tempTextArea.style.left = '-1000px';
  document.body.insertBefore(tempTextArea, document.body.firstChild);
  tempTextArea.focus();
  document.body.removeChild(tempTextArea);
}
```

The EventBus option would also work.

```js
SirTrevor.EventBus.on('block:create:existing', (block) => {
  // do something
});
```

**Comment by andrewsprinz** [comment]
Nice, thanks, managed to hack it with the below until I can have a proper look. Not pretty, but you know me :)

`
SirTrevor.EventBus.on('block:create:existing',function(block){ 
    $('html,body').scrollTop(0); 
});
`

**Comment by raffij** [comment]
Similar thoughts. Both are hacks.

**Comment by coneybeare** [comment]
Yikes, these are all kinda iffy

**Comment by SgtOddball** [comment]
If it looks stupid but it works, it's not stupid....... :grinning: 

(Murphy was a programmer)

**Comment by raffij** [comment]
@coneybeare we were looking for a simple fix for now. So in the near future can put a better consistent config in place without changing in the future.

**Comment by mdesantis** [comment]
ehm... what is the fix?? I'm trying in this way:

```coffeescript
$(document).on 'ready turbolinks:load', ->
  SirTrevor.config.language = 'it'
  
  SirTrevor.setDefaults(
    language: 'it'
    blockTypes: ['Text', 'Image', 'List', 'Heading', 'Quote']
    defaultType: 'Text'
    iconUrl: "<%= asset_path('sir-trevor/build/sir-trevor-icons.svg') %>"
    uploadUrl: "/admin/uploads/images"
    ajaxOptions: {
      headers: {
        'X-Requested-With': 'XMLHttpRequest'
        'X-CSRF-Token': Rails.csrfToken()
      }
      credentials: 'same-origin'
    }
  )

  $sirTrevorInstances = $('.js-sir-trevor-editor')

  for el in $sirTrevorInstances
    elConfig = $(el).data('sir-trevor-editor-config') ? {}
    config = { el: el }
    
    new (SirTrevor.Editor)($.extend({}, config, elConfig))

  if $sirTrevorInstances.length > 0
    # https://github.com/madebymany/sir-trevor-js/issues/349#issuecomment-218684054
    # Fixes this: https://github.com/madebymany/sir-trevor-js/issues/349
    SirTrevor.EventBus.on 'block:create:existing', (block) ->
      setTimeout () ->
        $('html, body').scrollTop(0)
      , 150
```

But no luck :cry:

**Comment by ilyaporopudas** [comment]
Fixed by #534
